### PR TITLE
Kelsha Darby

### PR DESCRIPTION
**Pull request by kelshadarby**
I was not able to get my details test to pass but it was passing in pry using the hash and the details method. I'm not sure if assert_equal was the wrong choice for this one.

### CODE REVIEW

File: lib/house.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class House
+
+  attr_reader :price, :address, :rooms
+
+  def initialize(price, address)
+    @price = price[1..-1].to_i
```
Comment (aziobrow) [inline lib/house.rb:6]: nice!

File: lib/house.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  def add_room(room)
+    @rooms << room
+  end
+
+  def above_market_average?
+    if @price > 500000
+      true
+    else
+      false
+    end
```
Comment (aziobrow) [inline lib/house.rb:20]: This is great.

It could also be written in a single line as:

`@price > 500000`, since that statement in and of itself returns a boolean value

File: lib/house.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def area
+    room_areas = []
+    @rooms.each do |room|
+      room_areas << room.area
+    end
+    room_areas.sum
```
Comment (aziobrow) [inline lib/house.rb:28]: This is great, and works well.  `Sum` is an enumerable of its own, and could have been used instead of `each. 

```
def area 
    @rooms.sum do |room|
         room.area
     end
end
```

it would iterate over each room and sum the area of each room to get you the total

File: lib/house.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def rooms_from_category(category)
+    @rooms.find_all do |room|
+      if room.category == category
+        room
+      end
+    end
```
Comment (aziobrow) [inline lib/house.rb:36]: Nice job choosing the best tool for the job! `find_all` is a great enumerable here. `find_all` only takes a boolean condition though, so the way this should actually  be written is as follows:

```
def rooms_from_category(category)
   @rooms.find_all do |room|
         room.category == category 
    end
end 
```

I think in this case, you got a little bit lucky, and it works anyway, but you can get yourself into trouble, setting yourself up for false positives, by putting anything other than a conditional into a `find_all`

File: lib/house.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def details
+    details = {"price" => @price, "address" => @address}
```
Comment (aziobrow) [inline lib/house.rb:40]: 👍

File: lib/room.rb [hunk 6]
```diff
@@ -1,12 +1,27 @@
...
+class Room
+
+  attr_reader :category
 
   def initialize(category, length, width)
     @category = category
-    @length = 10
+    @length = length
     @width = width
+    @paint = nil
```
Comment (aziobrow) [inline lib/room.rb:9]: we'd probably set this to false (i.e., when a house is very first created, it is unpainted).  This is logical, and additionally, it makes it a little easier since you don't need the conditional in your `is_painted?` method.

File: lib/room.rb [hunk 7]
```diff
@@ -1,12 +1,27 @@
...
-    @length * @width
+    @length * @width.to_i
+  end
+
+  def is_painted?
+    if @paint == nil
+      false
+    else
+      true
+    end
```
Comment (aziobrow) [inline lib/room.rb:21]: If you default `@painted` to false, this whole method can just be:

```
def is_painted?
    @painted 
end
```

File: lib/room.rb [hunk 8]
```diff
@@ -1,12 +1,27 @@
...
+  end
+
+  def paint
+    @paint = true
```
Comment (aziobrow) [inline lib/room.rb:25]: 👍

File: test/house_test.rb [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+require './lib/room'
+require './lib/house'
+
+class HouseTest < Minitest::Test
+
+  def test_it_exists_and_has_attributes
+    house = House.new("$400000", "123 sugar lane")
+
+    assert_instance_of House, house
+    assert_equal 400000, house.price
```
Comment (aziobrow) [inline test/house_test.rb:12]: nice catch here

File: test/house_test.rb [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+  def test_it_can_have_rooms
+    house = House.new("$400000", "123 sugar lane")
+    room_1 = Room.new(:bedroom, 10, '13')
+
+    assert_instance_of Room, room_1
+
+    room_2 = Room.new(:bedroom, 11, '15')
+
+    assert_instance_of Room, room_2
```
Comment (aziobrow) [inline test/house_test.rb:31]: It doesn't hurt to do two different rooms, but unless there's something substantially different about these two rooms, it probably doesn't do you any good, either, so we can just test one.

File: test/house_test.rb [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    room_4 = Room.new(:basement, 30, '41')
+    house.add_room(room_1)
+    house.add_room(room_2)
+
+    assert_equal [room_1, room_2], house.rooms
+
+    house.add_room(room_3)
+    house.add_room(room_4)
+
+    assert_equal [room_1, room_2, room_3, room_4], house.rooms
```
Comment (aziobrow) [inline test/house_test.rb:48]: Same here-- I like the thoroughness of your thought process here, but unless there's something different about adding 4 rooms instead of 2, there's probably no reason to do the extra two rooms/assertions here.

File: test/house_test.rb [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    room_3 = Room.new(:living_room, 25, '15')
+    room_4 = Room.new(:basement, 30, '41')
+    house.add_room(room_1)
+    house.add_room(room_2)
+    house.add_room(room_3)
+    house.add_room(room_4)
+
+    assert_equal [room_1, room_2], house.rooms_from_category(:bedroom)
+    assert_equal [room_4], house.rooms_from_category(:basement)
+    assert_equal [room_3], house.rooms_from_category(:living_room)
```
Comment (aziobrow) [inline test/house_test.rb:64]: doing more than one category here is a good idea though! does it work if you pass in a category that doesn't exist? (it should, but that could be a good thing to test! although for a timed exercise, i wouldn't worry about edge case testing too much)

File: test/house_test.rb [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    room_1 = Room.new(:bedroom, 10, '13')
+    room_2 = Room.new(:bedroom, 11, '15')
+    room_3 = Room.new(:living_room, 25, '15')
+    room_4 = Room.new(:basement, 30, '41')
+    house.add_room(room_1)
+    house.add_room(room_2)
+    house.add_room(room_3)
+    house.add_room(room_4)
+
+    assert_equal 1900, house.area
```
Comment (aziobrow) [inline test/house_test.rb:78]: 👍

File: test/house_test.rb [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+  end
+
+  def test_it_can_be_above_market_average?
+    house = House.new("$400000", "123 sugar lane")
+
+    assert_equal false, house.above_market_average?
```
Comment (aziobrow) [inline test/house_test.rb:84]: in this case, probably a good idea to create another house that is above market average and make sure that the other possibility works as well.

File: test/room_test.rb [hunk 15]
```diff
@@ -9,16 +9,24 @@ def test_it_exists
...
   end
+
+  def test_it_can_tell_if_room_can_be_painted
+    room = Room.new(:bedroom, 10, '13')
+
+    assert_equal false, room.is_painted?
+
+    room.paint
+
+    assert_equal true, room.is_painted?
```
Comment (aziobrow) [inline test/room_test.rb:30]: nicely done-- good progression here

**Review by aziobrow — COMMENTED** [review]
Great work! You got through iteration 3 accurately and in time, which is exactly what we're hoping for.  

Looks like you didn't get to the three methods in iteration 4, which is completely fine.  Loved your attention to detail-- you did a great job of both implementation and following the interaction pattern exactly.  I left some inline comments with a bit more detail. 

Let me know if you have any questions!

File: lib/house.rb (hunk 3 above)
Comment (kelshadarby) [reply lib/house.rb:28]: Woahhh! I didn't know sum was also an enumerable! Thanks for calling that out!
